### DISCUSSION TITLE
Tauri, wasm and wasi

### DISCUSSION THREAD

**Discussion by dev-mash** [tauri-apps/tauri#9521] [Q&A] [answered]
Hi,

I just started  with rust and tauri. I also came across webassembly (wasm) and webassembly system interace (wasi) which is also new to me.

From the tauri architecture I understood that WRY (os webview abstraction) and TAO is used to handle window creation and rendering (webview). I'm curious, is tauri using wasm/wasi to enable the communication between javascript and rust or is this not required?  Can you point me to more details how the communication between javascript and rust is done?

Best regards,
Manuel

**Comment by FabianLars** [comment, marked as answer]
> From the tauri architecture I understood that WRY (os webview abstraction) and TAO is used to handle window creation and rendering (webview).

Almost. Tao is _only_ the windowing library. Tao alone will only get you an empty white window. Wry is indeed the abstraction layer for the os webview, and the webviews themselves are what render your website into the Tao window.

>  is tauri using wasm/wasi to enable the communication between javascript and rust or is this not required?

There is no WASM involved anywhere in Tauri unless you chose a frontend framework that uses WASM (for example leptos or yew) but even then the IPC will still run over the javascript implementations.

WASI is not involved and not supported. (unless you bundle a runtime yourself alongside your tauri app and run another app on that runtime, but let's not go there lol)

> Can you point me to more details how the communication between javascript and rust is done?

We don't have this throughly document (though would love to at some point) but in v1 we simply use the webview's postMessage apis to send json string messages (don't ask me how they work internally 😅 ) and in v2 we additionally use http request (caught with native webview apis and without a webserver before they actually reach the network) to be able to handle larger ipc payloads.

**Reply by dev-mash** [reply]
Thanks for this explanation!
